io-sim-classes/src/Control/Monad/Class/MonadSTM/Strict.hs

```diff
@@ -131,9 +150,11 @@ readTVarIO :: MonadSTM m => StrictTVar m a -> m a
 readTVarIO StrictTVar { tvar } = Lazy.readTVarIO tvar
 
 writeTVar :: (MonadSTM m, HasCallStack) => StrictTVar m a -> a -> STM m ()
-writeTVar StrictTVar { tvar, invariant } !a =
-    checkInvariant (invariant a) $
-    Lazy.writeTVar tvar a
+writeTVar v !a =
+#if CHECK_TVAR_INVARIANT
```

Comment:
This seems unnecessary, since the definition of `checkInvariant` already uses CPP to make it a no op without the flag.

Reply:
Ah, I misread the diff here too. It's about `invariant` being in scope or not.

This is fine as is. But I'm thinking we could consolidate the CPP into a single definition of 

```
getInvariant :: StrictTVar -> a -> Maybe String
```

And then use `getInvariant` to define both `castStrictTVar` and `writeTVar` without CPP.

Reply:
We need `getInvariant` too and `setInvariant` to consolidate the CPP pragrams. Check 11dc98633.

Reply:
It might be nicer to use `mkStrictTVar :: (a -> Maybe String) -> LazyTVar m a -> StrictTVar m a`. Unlike `setInvariant`, that wouldn't require partial record expressions.

At this point, with `getInvariant` and `mkStrictTVar`, we've created an abstract interface that handles both values of the CPP -D variable. The interface looks like invariants are always present, but we can ignore them and/or use `const Nothing` as a filler when their disabled/irrelevant.

Reply:
> It might be nicer to use mkStrictTVar :: (a -> Maybe String) -> LazyTVar m a -> StrictTVar m a. Unlike setInvariant, that wouldn't require partial record expressions.

Yes I realised that `setInvariant` is partial, while `mkStirctTVar` is not.